keps/sig-storage/5030-attach-limit-autoscaler/README.md

```diff
@@ -0,0 +1,717 @@
+
+### Scaling a node-group that already has one or more nodes.
+
+1. We propose a similar label as GPULabel added to the node that is supposed to come up with a CSI driver. This would ensure that, nodes which are supposed to have a certain CSI driver installed aren’t considered ready - https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L979 until CSI driver is installed there.
+
+However, we also propose that a node will be considered ready as soon as corresponding CSI driver is being reported as installed via corresponding CSINode object.
+
+A node which is ready  but does not have CSI driver installed within certain time limit will be considered as NotReady and removed from the cluster.
+
+
+2. We propose that, we add volume limits and installed CSI driver information to framework.NodeInfo objects. So -
+
+```
+type NodeInfo struct {
+....
+....
+csiDrivers map[string]*DriverAllocatable
+..
+}
+
+type DriverAllocatable struct {
+    Allocatable int32
+}
+```
+
+3. We propose that, when saving `ClusterState` , we capture and add `csiDrivers` information to all existing nodes.
+
+4. We propose that, when getting nodeInfosForGroups , the return nodeInfo map also contains csidriver information, which can be used later on for scheduling decisions.
+
+```
+nodeInfosForGroups, autoscalerError := a.processors.TemplateNodeInfoProvider.Process(autoscalingContext, readyNodes, daemonsets, a.taintConfig, currentTime)
+```
+
+Please note that, we will have to handle the case of scaling from 0, separately from
+scaling from 1, because in former case - no CSI volume limit information will be available
+If no node exists in a NodeGroup.
+
+5. We propose that, when deciding pods that should be considered for scaling nodes in podListProcessor.Process function, we update the hinting_simulator to consider CSI volume limits of existing nodes. This will allow cluster-autoscaler to exactly know, if all unschedulable pods will fit in the recently spun or currently running nodes.
+
+Making aforementioned changes should allow us to handle scaling of nodes from 1.
+
+### Scaling from zero
+
+Scaling from zero should work similar to scaling from 1, but the main problem is - we do not have NodeInfo which can tell us what would be the CSI attach limit on the node which is being spun up in a NodeGroup.
+
+We propose that we introduce similar annotation as CPU, Memory resources in cluster-api to process attach limits available on a node.
+
+We have to introduce similar mechanism in various cloudproviders which return Template objects to incorporate volume limits. This will allow us to handle the case of scaling from zero.
+
+
+## Kubernetes Scheduler change
+
+We also propose that, if given node is not reporting any installed CSI drivers, we do not schedule pods that need CSI volumes to that node.
+
+The proposed change is small and a draft PR is available here - https://github.com/kubernetes/kubernetes/pull/130702
```

Comment:
This PR does not reference the feature gate you mention here?

Will this work be gated by the same feature gate or are these two separate features?